### PR TITLE
[OTX] Recover Semi-SL for multi-class classification

### PR DESCRIPTION
**Pull request by JihwanEom**
### Summary
- Update classification dataset to handle Semi-SL data pipeline
- Add templates for Semi-SL
- Remove num_workers in templates (use configuration.yaml, unify with other tasks)
- Update inference task & train task to handle Semi-SL data pipeline and unlabeled dataset, respectively.
- Remove argument expectation in mmcv/runner.py to support ComposedDL

Related PR: https://github.com/openvinotoolkit/model_preparation_algorithm/pull/100
Please note PR will be merged after merging MPA PR

### CODE REVIEW

**Review by harimkang — COMMENTED** [review]
Most of the reviews have been left in PR on the MPA side, and it seems necessary to receive an unlabeled batch size or ratio from OTX. Also, since there is no test code, please add it.

**Comment by JihwanEom** [comment]
> 

Thank you for your comments, I'll add tests after reflecting comments from MPA PR.

**Comment by eunwoosh** [comment]
For HPO search space, lr can be set as default_lr * 0.1 *~ default_lr * 10 (be careful that not exceeding maximum value) and bs can be set as default_bs * 0.5 ~ default_bs * 2 generally.

File: otx/algorithms/common/adapters/mmcv/runner.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def stop(self) -> bool:
         return broadcast_obj[0]
 
     @check_input_parameters_type()
-    def train(self, data_loader: DataLoader, **kwargs):
+    def train(self, data_loader, **kwargs):
```
Comment (eunwoosh) [inline otx/algorithms/common/adapters/mmcv/runner.py]: why do you remove type hinting. Can other type dataloader be passed to this function?

File: otx/algorithms/classification/tasks/train.py [hunk 2]
```diff
@@ -156,6 +156,10 @@ def _init_train_data_cfg(self, dataset: DatasetEntity):
                     otx_dataset=dataset.get_subset(Subset.VALIDATION),
                     labels=self._labels,
                 ),
+                unlabeled=ConfigDict(
+                    otx_dataset=dataset.get_subset(Subset.UNLABELED),
+                    labels=self._labels,
```
Comment (eunwoosh) [inline otx/algorithms/classification/tasks/train.py]: It's just my curiosity. unlabled should need to get lables?

File: otx/algorithms/classification/tasks/inference.py [hunk 3]
```diff
@@ -435,7 +436,11 @@ def patch_color_conversion(pipeline):
             for pipeline_step in cfg.pipeline:
                 if subset == "train" and pipeline_step.type == "Collect":
                     pipeline_step = get_meta_keys(pipeline_step)
-            patch_color_conversion(cfg.pipeline)
+
+            if isinstance(cfg.pipeline, ConfigDict):
+                cfg.pipeline = {key: patch_color_conversion(val) for key, val in cfg.pipeline.items()}
```
Comment (eunwoosh) [inline otx/algorithms/classification/tasks/inference.py]: if you want patch_color_conversion() to all values in cfg.pipeline, then how about to use code as below?
`for val in cfg.pipeline.values(): patch_color_conversion(val)`
and remove `return` in patch_color_conversion().

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Description.
```
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml:1]: template_semisl.yaml is more appropriate

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+model_template_id: Custom_Image_Classification_EfficinetNet-B0
```
Comment (jaegukhyun) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml:2]: For test, I want to suggest to change model_template_id to SemiSL_Image_*

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
Generally it looks good to me, but I want to move the way to handle multiple pipeline for semisl from otx to mpa. I recommend to look at BranchImage pipeline in semisl-det

File: otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/semisl/template_experimental.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+application: ~
+
+# Algo backend.
+framework: OTEClassification v1.2.3
+
+# Task implementations.
+entrypoints:
+  base: otx.algorithms.classification.tasks.ClassificationTrainTask
+  openvino: otx.algorithms.classification.tasks.ClassificationOpenVINOTask
+  nncf: otx.algorithms.classification.tasks.nncf.OTXClassificationNNCFTask
```
Comment (sovrasov) [inline otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/semisl/template_experimental.yaml:17]: Do we have a working nncf implementation for the semi-sl right now? Will optimize work with this template if we omit unlabeled data?

File: otx/algorithms/classification/adapters/mmcls/data/datasets.py [hunk 7]
```diff
@@ -96,9 +97,16 @@ def __getitem__(self, index):
             ignored_labels=ignored_labels,
         )
 
-        if self.pipeline is None:
-            return data_info
-        return self.pipeline(data_info)
+        data_infos = [copy.deepcopy(data_info) for _ in range(self.num_pipes)]
```
Comment (wonjuleee) [inline otx/algorithms/classification/adapters/mmcls/data/datasets.py]: do we need to copy `data_info` `self.num_pipes` times?

File: otx/algorithms/classification/adapters/mmcls/data/datasets.py [hunk 8]
```diff
@@ -96,9 +97,16 @@ def __getitem__(self, index):
...
+        if isinstance(self.pipeline, dict):
+            results = {}
+            for i, k in enumerate(self.pipeline.keys()):
+                results[k] = self.pipeline[k](data_infos[i])
```
Comment (wonjuleee) [inline otx/algorithms/classification/adapters/mmcls/data/datasets.py]: Could you please test the operation below?
for k, v in self.pipeline.items():
    results[k] = v(data_info)

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template.yaml [hunk 9]
```diff
@@ -29,8 +29,6 @@ hyper_parameters:
       batch_size:
         default_value: 64
         auto_hpo_state: POSSIBLE
-      num_workers:
-        default_value: 0
```
Comment (wonjuleee) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template.yaml:34]: can I ask why you remove this?

File: otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template.yaml [hunk 10]
```diff
@@ -29,8 +29,6 @@ hyper_parameters:
       batch_size:
         default_value: 64
         auto_hpo_state: POSSIBLE
-      num_workers:
-        default_value: 0
```
Comment (wonjuleee) [inline otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/template.yaml:34]: same as above.

File: otx/algorithms/common/adapters/mmcv/runner.py (hunk 1 above)
Comment (wonjuleee) [inline otx/algorithms/common/adapters/mmcv/runner.py]: Are you considering multiple types of data_loader could be fed into train function?

File: otx/algorithms/classification/tasks/train.py (hunk 2 above)
Comment (JihwanEom) [reply otx/algorithms/classification/tasks/train.py]: Updated to set as `None`, thank you for comment :)

File: otx/algorithms/classification/tasks/inference.py (hunk 3 above)
Comment (JihwanEom) [reply otx/algorithms/classification/tasks/inference.py]: Removed these lines by using BranchImage. Thank you for suggestion.

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml (hunk 4 above)
Comment (JihwanEom) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml:1]: Will be reorganized to use one template.

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml (hunk 5 above)
Comment (JihwanEom) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/semisl/template_experimental.yaml:2]: Same with above

File: otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/semisl/template_experimental.yaml (hunk 6 above)
Comment (JihwanEom) [reply otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/semisl/template_experimental.yaml:17]: It should be discussed for all semi-sl tasks. @supersoob @jaegukhyun  Thank you for opinion.

File: otx/algorithms/classification/adapters/mmcls/data/datasets.py (hunk 7 above)
Comment (JihwanEom) [reply otx/algorithms/classification/adapters/mmcls/data/datasets.py]: Removed these lines by using BranchImage. Thank you for comment.

File: otx/algorithms/classification/adapters/mmcls/data/datasets.py (hunk 8 above)
Comment (JihwanEom) [reply otx/algorithms/classification/adapters/mmcls/data/datasets.py]: Same as above.

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template.yaml (hunk 9 above)
Comment (JihwanEom) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/template.yaml:34]: Detection, segmentation controls num_workers by updating in configuration.yaml. I've aligned with them. (Also verified it works)

File: otx/algorithms/common/adapters/mmcv/runner.py (hunk 1 above)
Comment (JihwanEom) [reply otx/algorithms/common/adapters/mmcv/runner.py]: It is also resolved by using unlabeled_hook. Previous implementation was building ComposedDL before the training, I modified to change the dataloader dynamically from torch.DataLoader to ComposedDL.

**Comment by JihwanEom** [comment]
Close by #1473

**Pull request closed without merging**